### ISSUE TITLE
Separate cache timeout for profiles

### ISSUE THREAD

**Issue by renevangsgaardjp** [pac4j/play-pac4j#378]
It would be nice if timeout values for profiles in the cache could differ. The motivation is that some profiles/sessions should live for a long time (OpenID offline_access), and other (Anonymous profiles) should live for a much shorter time period.

My current idea is to implement a `org.pac4j.play.store.PlayCacheStore`, but at that point in the flow, data is just strings, which seem like a terrible hack. Do you have any ideas on how to do this?

After implementing it in a Play project, I will be imlementing it in a Spring Web MVC app with Pac4j as well.

**Comment by leleuj** [comment]
You can of course create your own implementation of the `PlayCacheSessionStore` or of the `SessionStore` more generally, but this can be some work and lead to some maintenance work later on.
You can also override the profiles types with new ones and change the `isExpired` method logic.

**Comment by renevangsgaardjp** [comment]
I do not think that using isExpired will meet my end goal: allowing the cache implementation (in my case Redis) to delete entries before Pac4j discovers the profile has timed out, if ever. If a low timeout is not set, the cache will be filled with a lot of anonymous profiles.

Another approach could be that Pac4j wouldn't need to store profile data - I guess it has a big impact on Pac4j and I really don't know if the idea is any good.

**Comment by leleuj** [comment]
The best is to have your own `SessionStore` implementation. This way, you can control timeout or custom logic.

**Comment by renevangsgaardjp** [comment]
Thank you for taking your time to respond.

This is currently how I have implemented it - can you see another/better way?

```
import org.slf4j.LoggerFactory
import play.cache.SyncCacheApi

import java.util.{Map => JMap}

class PlayCacheSessionStore(
    cache: SyncCacheApi,
    timeout: Int,
    anonTimeout: Int
) extends org.pac4j.play.store.PlayCacheSessionStore(cache) {
  store = new PlayCacheStore[String, JMap[String, Object]](
    cache,
    timeout,
    anonTimeout
  )
}

class PlayCacheStore[K, O](
    cache: SyncCacheApi,
    timeout: Int,
    anonTimeout: Int
) extends org.pac4j.play.store.PlayCacheStore[K, O](cache) {
  import PlayCacheStore._

  val logger = LoggerFactory.getLogger(getClass)

  // Setting it just to have the value set correctly, before the super class logs it.
  // The value is not really used, as timeout is controlled in this class.
  setTimeout(timeout)

  override def internalSet(key: K, value: O): Unit = {
    getCache.set(computeKey(key), value, computeTimeout(value))
  }

  protected def computeTimeout(value: O): Int = {
    if (isAnonymous(value)) anonTimeout else timeout
  }

  override def toString: String =
    super.toString + s" anonTimeout: ${anonTimeout} |"
}

object PlayCacheStore {

  val Pac4jUserProfilesKey = "pac4jUserProfiles"
  val Pac4jDefaultTKey = "DEFAULT"

  def isAnonymous(value: Any): Boolean = {
    value match {
      case map: JMap[_, _] =>
        map.get(Pac4jUserProfilesKey) match {
          case profileMap: JMap[_, _] =>
            profileMap.get(Pac4jDefaultKey) match {
              case _: AnonymousProfile => true
              case _                   => false
            }
          case _ => false
        }
      case _ => false
    }
  }
}
```

**Comment by leleuj** [comment]
It looks good to me.

**Comment by SoerenSilkjaer** [comment]
@leleuj I can you see you have been working on fixing the PlayCookieSessionStore. What did you actually fix? I remember not being able to use the cookie store earlier because any session was too large to store in a cookie. If it is indeed the case that it is now fixed, I suppose that would help with the issues @renevangsgaardjp describes. If all his sessions are saved in cookies, there is no redis to fill up with anonymous profiles.

**Comment by SoerenSilkjaer** [comment]
It would however require the release of pac4j v5. when do you expect to release this new version?

**Comment by leleuj** [comment]
Yes, this is the core issue, the size of the cookie.
I have done two things:
- changed pac4j to limit data: erased data are set to `null` instead of the empty string, the requested URL is now a string instead of a `FoundAction`
- reworked play-pac4j to have the `PlayCookieSessionStore` saved all data in a serialized/deflated/encrypted `HashMap` instead of separate (key, value) and remove `null` data.

5.0.0-RC1 is planned for next week and the final release for March/April.
